src/MonoMod.UnitTest/RuntimeDetour/ILCursorTest.cs

```diff
@@ -0,0 +1,183 @@
+                })
+            );
+        }
+
+        [Fact]
+        public void TestMoveIntoHandler() {
+            // Deliberately moves into the catch clause with filter
+            Assert.Equal("tryTESTwhenfinallyreturn",
+                TestTryCatchFinallyILEdit(c => {
+                    c.Goto(c.Body.ExceptionHandlers.Single(eh => eh.FilterStart != null).HandlerStart, MoveType.AfterLabel);
+                    EmitAppendTest(c);
+                })
+            );
+        }
+
+        [Fact]
+        public void TestMoveAfterHandlers() {
+            // After the finally block
+            Assert.Equal("trywhenfinallyTESTreturn",
+                TestTryCatchFinallyILEdit(c => {
+                    c.GotoNext(MoveType.AfterLabel,
+                        i => i.MatchLdloc(0),
+                        i => i.MatchLdstr("return"));
+                    EmitAppendTest(c);
+                })
+            );
+        }
+
+        [Fact]
+        public void TestRemoveMaintainsLabelPosition() {
+            // throw statement inserted into try block
+            Assert.Equal("catchfinallyreturn",
+                TestTryCatchFinallyILEdit(c => {
+                    c.GotoNext(MoveType.Before, i => i.MatchLdloc(0), i => i.MatchLdstr("try"));
+                    c.MoveAfterLabels();
+                    c.RemoveRange(4);
+                    EmitThrowEx(c);
+                })
+            );
+
+            // throw statement inserted before the try block
+            Assert.Throws<Exception>(() => {
+                TestTryCatchFinallyILEdit(c => {
+                    c.GotoNext(MoveType.Before, i => i.MatchLdloc(0), i => i.MatchLdstr("try"));
+                    c.MoveAfterLabels(intoEHRanges: false);
+                    c.RemoveRange(4);
+                    EmitThrowEx(c);
+                });
+            });
+
+            // throw statement inserted in unreachable if statement
+            Assert.Equal("whenfinallyreturn",
+                TestTryCatchFinallyILEdit(c => {
+                    c.GotoNext(MoveType.Before, i => i.MatchLdloc(0), i => i.MatchLdstr("try"));
+                    c.RemoveRange(4);
+                    EmitThrowEx(c);
+                })
+            );
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        internal static string TryCatchFinally() {
+            var sb = new StringBuilder();
+            if (sb.GetType() != typeof(StringBuilder)) {
+                sb.Append("unreachable");
+            }
+            try {
+                sb.Append("try");
+                throw new ArgumentException("test");
+            }
+            catch (Exception ex) when (ex is ArgumentException) {
+                sb.Append("when");
+            }
+            catch {
+                sb.Append("catch");
+            }
+            finally {
+                sb.Append("finally");
+            }
+            sb.Append("return");
+            return sb.ToString();
+        }
```

Comment:
This approach does worry me slightly, because it's dependent on how the compiler chooses to emit IL for this code. I don't expect it to actually be a problem, however I worry that over time these tests will start to fail over compiler version changes.

I'm fine with accepting this as-is, though, for now.

Reply:
The alternative is building the IL for this kind of test from scratch, which has its own drawbacks.